File: momentum_setups.py
# Read stocks
import yfinance as yf

# For plotting
import plotly.graph_objects as go
from plotly.subplots import make_subplots
import plotly.io as pio
pio.renderers.default = "browser"

# Data & TA
import pandas as pd
import numpy as np
import talib as ta

# Config
from jproperties import Properties
import datetime

# -------------------------------------------------
# Helper: load tickers from JSON (kept from original)
# -------------------------------------------------
def getTickersFromFile(filename):
    tickers = []
    with open(filename, 'r') as file:
        data = file.read()
        lines = data.splitlines()
        for line in lines:
            if '"ticker":' in line:
                ticker = line.split('"ticker":')[1].split('"')[1]
                tickers.append(ticker)
    return tickers

# -------------------------------------------------
# Dates & Config
# -------------------------------------------------
today = datetime.date.today().strftime("%Y-%m-%d")
twelve_months_ago = (datetime.date.today() - datetime.timedelta(days=365)).strftime("%Y-%m-%d")
five_years_ago = (datetime.date.today() - datetime.timedelta(days=1825)).strftime("%Y-%m-%d")

configs = Properties()
with open('config.properties', 'rb') as config_file:
    configs.load(config_file)

TICKER   = configs.get('TICKER').data
TEMPLATE = configs.get('TEMPLATE').data
TICKERS  = configs.get('TICKERS').data
NYSE     = configs.get('NYSE').data
NASDAQ    = configs.get('NASDAQ1B').data
TEST = configs.get('TEST').data
start = five_years_ago
tick_list = TICKERS.split()
# NYSE_list = NYSE.split()
# tick_list = NASDAQ.split()
tick_list = TEST.split()

# -------------------------------------------------
# Options liquidity filter
# -------------------------------------------------
def has_liquid_options(symbol, min_oi=3500, max_spread_pct=0.05):
    """
    Strict options liquidity filter.
    - Minimum Open Interest: 500
    - Maximum bid-ask spread: 5% of mid
    """
    t = yf.Ticker(symbol)
    try:
        if not t.options:
            return False

        exp = t.options[0]  # nearest expiration
        chain = t.option_chain(exp)
        calls = chain.calls.copy()
        puts  = chain.puts.copy()

        calls['openInterest'] = pd.to_numeric(calls['openInterest'], errors='coerce').fillna(0)
        puts['openInterest']  = pd.to_numeric(puts['openInterest'],  errors='coerce').fillna(0)

        max_call_oi = calls['openInterest'].max() if not calls.empty else 0
        max_put_oi  = puts['openInterest'].max()  if not puts.empty  else 0

        if max(max_call_oi, max_put_oi) < min_oi:
            return False

        # Force single best contract
        if max_call_oi >= max_put_oi:
            best = calls[calls['openInterest'] == max_call_oi].iloc[0]
        else:
            best = puts[puts['openInterest'] == max_put_oi].iloc[0]

        bid = float(best['bid']) if pd.notna(best['bid']) else 0.0
        ask = float(best['ask']) if pd.notna(best['ask']) else 0.0
        mid = (bid + ask) / 2

        if mid <= 0:
            return False

        spread = (ask - bid) / mid
        return spread <= max_spread_pct   # now 5%

    except Exception:
        return False


# -------------------------------------------------
# Main momentum-continuation scanner (day-trade / open focused)
# -------------------------------------------------

def analyze_momentum_candidate(symbol, start, debug=False):
    """
    True momentum continuation scorer for open / day-trade setups.
    Priorities (in order):
    1. Closed near previous-day high
    2. High relative volume (especially last day + 5d)
    3. Already trending upward (price > SMAs, +DI > -DI, ADX rising or strong)
    4. Relative strength vs SPY
    Hard filters kept: liquid enough, options optional, price sanity.
    """
    try:
        t = yf.Ticker(symbol.replace('.', '-'))
        # Faster window is fine for day-trade signals
        df = t.history(start=start, auto_adjust=True)[['Open', 'High', 'Low', 'Close', 'Volume']]
        if len(df) < 120:
            return None

        latest_price = df['Close'].iloc[-1]
        if latest_price > 350 or latest_price < 5:   # relaxed upper, still avoid micro-junk
            return None

        # Indicators (safe)
        try:
            df['SMA20']    = ta.SMA(df['Close'], 20)
            df['SMA50']    = ta.SMA(df['Close'], 50)
            df['SMA200']   = ta.SMA(df['Close'], 200)
            df['RSI']      = ta.RSI(df['Close'], 14)
            df['ATR']      = ta.ATR(df['High'], df['Low'], df['Close'], 14)
            df['ADX']      = ta.ADX(df['High'], df['Low'], df['Close'], 14)
            df['PLUS_DI']  = ta.PLUS_DI(df['High'], df['Low'], df['Close'], 14)
            df['MINUS_DI'] = ta.MINUS_DI(df['High'], df['Low'], df['Close'], 14)

            macd, macdsignal, macdhist = ta.MACD(df['Close'], 12, 26, 9)
            df['MACD_hist'] = macdhist
        except Exception:
            return None

        df = df.dropna(subset=['ADX', 'PLUS_DI', 'MINUS_DI', 'RSI', 'ATR', 'SMA20', 'SMA50'])
        if len(df) < 60:
            return None

        latest = df.iloc[-1]
        prev   = df.iloc[-2]
        prev2  = df.iloc[-3] if len(df) > 2 else prev

        # Liquidity
        dollar_vol_20d = (df['Close'] * df['Volume']).iloc[-20:].mean()
        if dollar_vol_20d < 15_000_000:          # slightly higher bar
            return None

        options_ok = has_liquid_options(symbol, min_oi=250, max_spread_pct=0.15)

        # Relative Strength vs SPY (3-month)
        spy = yf.Ticker('SPY').history(start=start, auto_adjust=True)['Close']
        common_idx = df.index.intersection(spy.index)
        if len(common_idx) < 70:
            return None
        stock_close = df['Close'].loc[common_idx]
        spy_close   = spy.loc[common_idx]
        lookback = 63
        stock_perf = stock_close.iloc[-1] / stock_close.iloc[-lookback] - 1
        spy_perf   = spy_close.iloc[-1] / spy_close.iloc[-lookback] - 1
        rs = (1 + stock_perf) / (1 + spy_perf) if (1 + spy_perf) != 0 else 0

        # -------------------------------------------------
        # KEY NEW METRICS for open momentum
        # -------------------------------------------------
        # 1. How close did it close to the high? (lower % = stronger)
        pct_from_high = (latest['High'] - latest['Close']) / latest['Close'] * 100
        atr_from_high = (latest['High'] - latest['Close']) / latest['ATR'] if latest['ATR'] > 0 else 99

        # 2. Relative volume
        vol_20 = df['Volume'].iloc[-20:].mean()
        vol_5  = df['Volume'].iloc[-5:].mean()
        rel_vol_today = latest['Volume'] / vol_20 if vol_20 > 0 else 0
        rel_vol_5d    = vol_5 / vol_20 if vol_20 > 0 else 0

        # 3. Recent higher highs / higher closes
        higher_high = latest['High'] > prev['High'] > prev2['High']
        higher_close = latest['Close'] > prev['Close'] > prev2['Close']
        consecutive_up = (latest['Close'] > prev['Close']) and (prev['Close'] > prev2['Close'])

        # -------------------------------------------------
        # SCORING – heavily tilted to near-high + volume + trend
        # -------------------------------------------------
        score = 0.0

        # A. Proximity to previous-day high  (max 35)  ← biggest change
        if pct_from_high <= 0.3 or atr_from_high <= 0.25:
            score += 35
        elif pct_from_high <= 0.7 or atr_from_high <= 0.5:
            score += 28
        elif pct_from_high <= 1.2 or atr_from_high <= 0.8:
            score += 18
        elif pct_from_high <= 2.0:
            score += 8

        # B. Volume surge (max 30)
        if rel_vol_today >= 2.5 or rel_vol_5d >= 2.0:
            score += 30
        elif rel_vol_today >= 1.8 or rel_vol_5d >= 1.6:
            score += 24
        elif rel_vol_today >= 1.4 or rel_vol_5d >= 1.3:
            score += 16
        elif rel_vol_today >= 1.1:
            score += 8

        # C. Trend structure (max 20)
        if latest['Close'] > latest['SMA20'] > latest['SMA50'] > latest['SMA200']:
            score += 20
        elif latest['Close'] > latest['SMA20'] > latest['SMA50']:
            score += 14
        elif latest['Close'] > latest['SMA50']:
            score += 8
        elif latest['SMA50'] > latest['SMA200']:
            score += 3

        # D. Directional movement (max 15)
        plus_di  = latest['PLUS_DI']
        minus_di = latest['MINUS_DI']
        di_spread = plus_di - minus_di
        if plus_di > minus_di:
            if di_spread >= 12:   score += 15
            elif di_spread >= 8:  score += 11
            elif di_spread >= 4:  score += 7
            else:                 score += 3

        # E. ADX strength / rising (max 12)
        adx = latest['ADX']
        adx_rising = adx > prev['ADX']
        if adx >= 35 and adx_rising:   score += 12
        elif adx >= 30:                score += 9
        elif adx >= 25:                score += 6
        elif adx >= 20:                score += 3

        # F. Consecutive strength bonus (max 10)
        if higher_high and higher_close:
            score += 10
        elif consecutive_up:
            score += 6
        elif latest['Close'] > prev['Close']:
            score += 3

        # G. Relative Strength (max 8)
        if rs >= 1.20:   score += 8
        elif rs >= 1.10: score += 6
        elif rs >= 1.03: score += 4
        elif rs >= 0.98: score += 1

        # H. RSI – mild momentum bonus (no heavy penalty)
        rsi = latest['RSI']
        if 55 <= rsi <= 78:
            score += 5
        elif rsi > 78:
            score += 2          # still allow parabolic continuation
        elif rsi < 40:
            score -= 6

        # I. MACD confirmation (max 5)
        if latest['MACD_hist'] > prev['MACD_hist'] and latest['MACD_hist'] > 0:
            score += 5
        elif latest['MACD_hist'] > 0:
            score += 2

        # Options liquidity (small)
        if options_ok:
            score += 2
        else:
            score -= 3

        # Risk:Reward (for reference only)
        atr = latest['ATR']
        entry = latest['Close']
        stop = entry - 1.2 * atr          # tighter for day trades
        target = entry + 2.5 * atr
        risk_reward = (target - entry) / (entry - stop) if (entry - stop) > 0 else 0

        return {
            'symbol': symbol,
            'score': round(score, 1),
            'close': round(entry, 2),
            'pct_from_high': round(pct_from_high, 2),
            'atr_from_high': round(atr_from_high, 2),
            'rel_vol_today': round(rel_vol_today, 2),
            'rel_vol_5d': round(rel_vol_5d, 2),
            'rsi': round(rsi, 1),
            'adx': round(adx, 1),
            'plus_di': round(plus_di, 1),
            'rs_3m': round(rs, 2),
            'atr': round(atr, 2),
            'stop': round(stop, 2),
            'target': round(target, 2),
            'risk_reward': round(risk_reward, 2),
            'dollar_vol_20d': int(dollar_vol_20d),
            'options_ok': bool(options_ok),
            'higher_high': bool(higher_high),
            'consecutive_up': bool(consecutive_up)
        }

    except Exception as e:
        if debug:
            print(f"{symbol}: error → {e}")
        return None

print("Scoring all tickers for open momentum...\n")
results = []
start = twelve_months_ago


for symbol in tick_list:
    result = analyze_momentum_candidate(symbol, start)
    if result:
        results.append(result)

top20 = sorted(results, key=lambda x: x['score'], reverse=True)[:20]

print("\n===== TOP 20 – Momentum Continuation (Near High + Volume + Trend) =====\n")
for i, r in enumerate(top20, 1):
    print(f"{i:2}. {r['symbol']:6} | Score: {r['score']:5.1f} | "
          f"Price: ${r['close']:6.2f} | FromHigh: {r['pct_from_high']:4.1f}% | "
          f"Vol: {r['rel_vol_today']:.1f}x | RSI: {r['rsi']:4.1f} | "
          f"ADX: {r['adx']:4.1f} | RR: {r['risk_reward']:.1f}")

# create json file with results
import json
datestamptext = datetime.datetime.now().strftime("%Y-%m-%d")
with open(f'momentum_results_{datestamptext}.json', 'w') as f:
    json.dump(top20, f, indent=4)

    

File: momentum_setups.mjs
import fs from 'fs/promises';
import path from 'path';
import PropertiesReader from 'properties-reader';
import yahooFinanceModule from 'yahoo-finance2';
const rawYahooFinance = yahooFinanceModule.default || yahooFinanceModule;
let yahooFinance = null;
try{
  // v4 exports a class that needs instantiation: `new YahooFinance()`
  // suppress one-time notices
  yahooFinance = (typeof rawYahooFinance === 'function') ? new rawYahooFinance({ suppressNotices: ['ripHistorical', 'yahooSurvey'] }) : rawYahooFinance;
}catch(e){
  // fallback to raw export
  yahooFinance = rawYahooFinance;
}
import { SMA, RSI, ATR, MACD, ADX } from 'technicalindicators';

const root = process.cwd();

function readConfig(file='config.properties'){
  try{
    const props = PropertiesReader(file);
    return {
      TICKER: props.get('TICKER'),
      TEMPLATE: props.get('TEMPLATE'),
      TICKERS: props.get('TICKERS'),
      NYSE: props.get('NYSE'),
      NASDAQ1B: props.get('NASDAQ1B'),
      TEST: props.get('TEST'),
      RISK_MODE: props.get('RISK_MODE')
    };
  }catch(e){
    console.error('Failed to read config.properties', e); return {};
  }
}

function formatDate(d){ return d.toISOString().slice(0,10); }
function tomorrow(){ const d = new Date(); d.setDate(d.getDate()+1); return formatDate(d); }

function normalizeRiskMode(cfg){
  const raw = process.env.RISK_MODE ?? cfg.RISK_MODE ?? 'normal';
  const mode = String(raw).trim().toLowerCase();
  return (mode === 'risk-off' || mode === 'riskoff' || mode === 'defensive') ? 'risk-off' : 'normal';
}

async function hasLiquidOptions(symbol, min_oi=250, max_spread_pct=0.25, min_volume=100){
  try{
    const opt = await yahooFinance.options(symbol);
    if(!opt || !opt.options || !opt.options.length) return false;
    const chain = opt.options[0];
    const price = opt.quote?.regularMarketPrice || opt.quote?.price || 0;
    const calls = chain.calls || [];
    const puts  = chain.puts  || [];
    const allContracts = [...calls, ...puts];
    if(!allContracts.length) return false;

    // Include any contract with a strike within 10% of current price (either direction)
    const otm_pct = 0.10;
    const lo = price * (1 - otm_pct);
    const hi = price * (1 + otm_pct);

    // Check if any contract within that band qualifies.
    // Primary: OI + bid/ask spread.
    // Fallback: if Yahoo returns zero bid/ask snapshots, allow strong OI+volume.
    return allContracts.some(c => {
      const strike = Number(c.strike);
      if(price > 0 && (strike < lo || strike > hi)) return false;
      const oi = Number(c.openInterest) || 0;
      if(oi < min_oi) return false;
      const bid = Number(c.bid) || 0;
      const ask = Number(c.ask) || 0;
      const vol = Number(c.volume) || 0;
      const mid = (bid + ask) / 2;
      if(mid > 0){
        return (ask - bid) / mid <= max_spread_pct;
      }
      // Data-quality fallback for delayed/zeroed bid-ask snapshots.
      return vol >= min_volume;
    });
  }catch(e){ return false; }
}

async function analyze(symbol, periodStart, riskMode='normal'){
  try{
    // chart() in yahoo-finance2 v4 returns { meta, quotes, events }
    // where quotes is an array of { date, open, high, low, close, volume }
    const sym = symbol.replace('.', '-');
    const chartData = await yahooFinance.chart(sym, { period1: periodStart, period2: tomorrow(), interval: '1d' });
    if(!chartData || !Array.isArray(chartData.quotes) || chartData.quotes.length === 0) return null;
    const hist = chartData.quotes.filter(h => h.close != null);
    if(!Array.isArray(hist) || hist.length < 120) return null;
    const close = hist.map(h=>h.close);
    const high = hist.map(h=>h.high);
    const low = hist.map(h=>h.low);
    const open = hist.map(h=>h.open);
    const volume = hist.map(h=>h.volume);

    // indicators
    const sma20 = SMA.calculate({ period:20, values: close });
    const sma50 = SMA.calculate({ period:50, values: close });
    const sma200 = SMA.calculate({ period:200, values: close });
    const rsi = RSI.calculate({ period:14, values: close });
    const atr = ATR.calculate({ period:14, high, low, close });
    let adx = [];
    try{ adx = ADX.calculate({ period:14, high, low, close }); }catch(e){ /* ignore */ }
    const macd = MACD.calculate({ values: close, fastPeriod:12, slowPeriod:26, signalPeriod:9, SimpleMAOscillator:false, SimpleMASignal:false });

    // align indexes to the latest
    const lastIdx = close.length - 1;
    const latest = {
      Close: close[lastIdx],
      High: high[lastIdx],
      Low: low[lastIdx],
      ATR: atr.length ? atr[atr.length-1] : 0,
      RSI: rsi.length ? rsi[rsi.length-1] : 0,
      ADX: adx.length ? adx[adx.length-1].adx || adx[adx.length-1] : (adx.length?adx[adx.length-1]:0),
      PLUS_DI: adx.length ? adx[adx.length-1].pdi || 0 : 0,
      MINUS_DI: adx.length ? adx[adx.length-1].mdi || 0 : 0,
      MACD_hist: macd.length ? macd[macd.length-1].histogram || (macd[macd.length-1].hist) || 0 : 0
    };

    // risk checks – no upper bound; skip penny stocks only
    if(latest.Close < 3) return null;
    if(latest.Close > 500) return null;
    // liquidity: dollar vol 20d
    const closes = close.slice(-20); const vols = volume.slice(-20);
    const dollarVol = closes.reduce((s,c,idx)=>s + c * vols[idx], 0) / Math.min(20, closes.length);
    if(dollarVol < 15_000_000) return null;

    const options_ok = await hasLiquidOptions(symbol, 250, 0.25, 100);

    // RS vs SPY via chart()
    const spyData = await yahooFinance.chart('SPY', { period1: periodStart, period2: tomorrow(), interval: '1d' });
    const spy_close = (spyData?.quotes || []).filter(q=>q.close!=null).map(q=>q.close);
    const commonLen = Math.min(spy_close.length, close.length);
    const lookback = Math.min(63, commonLen-1);
    const stock_perf = close[close.length-1] / close[close.length-1-lookback] - 1;
    const spy_perf = spy_close.length >= lookback+1 ? spy_close[spy_close.length-1] / spy_close[spy_close.length-1-lookback] - 1 : 0;
    const rs = spy_perf !== -1 ? (1+stock_perf)/(1+spy_perf) : 1;

    // compute pct from high and atr_from_high
    const pct_from_high = (latest.High - latest.Close)/latest.Close * 100;
    const atr_from_high = latest.ATR > 0 ? (latest.High - latest.Close)/latest.ATR : 99;

    // relative vol
    const vol20 = volume.slice(-20).reduce((s,v)=>s+v,0)/20;
    const vol5 = volume.slice(-5).reduce((s,v)=>s+v,0)/Math.min(5, volume.slice(-5).length);
    const rel_vol_today = volume[volume.length-1] / (vol20 || 1);
    const rel_vol_5d = vol5 / (vol20 || 1);

    // recent highs/closes
    const prev = { High: high[high.length-2], Close: close[close.length-2] };
    const prev2 = { High: high[high.length-3], Close: close[close.length-3] };
    const higher_high = latest.High > prev.High && prev.High > prev2.High;
    const higher_close = latest.Close > prev.Close && prev.Close > prev2.Close;
    const consecutive_up = latest.Close > prev.Close && prev.Close > prev2.Close;

    // scoring (mirror python)
    let score = 0.0;

    const riskOff = riskMode === 'risk-off';

    // risk-off guards to reduce late-stage chase names
    if(riskOff){
      if(rel_vol_5d < 1.15) score -= 10;
      if(pct_from_high > 1.4 || atr_from_high > 1.0) score -= 10;
    }

    if(pct_from_high <= 0.3 || atr_from_high <= 0.25) score += 35;
    else if(pct_from_high <= 0.7 || atr_from_high <= 0.5) score += 28;
    else if(pct_from_high <= 1.2 || atr_from_high <= 0.8) score += 18;
    else if(pct_from_high <= 2.0) score += 8;

    if(rel_vol_today >= 2.5 || rel_vol_5d >= 2.0) score += 30;
    else if(rel_vol_today >= 1.8 || rel_vol_5d >= 1.6) score += 24;
    else if(rel_vol_today >= 1.4 || rel_vol_5d >= 1.3) score += 16;
    else if(rel_vol_today >= 1.1) score += 8;

    // trend
    const sma20v = sma20[sma20.length-1];
    const sma50v = sma50[sma50.length-1];
    const sma200v = sma200[sma200.length-1];
    if(latest.Close > sma20v && sma20v > sma50v && sma50v > sma200v) score += 20;
    else if(latest.Close > sma20v && sma20v > sma50v) score += 14;
    else if(latest.Close > sma50v) score += 8;
    else if(sma50v > sma200v) score += 3;

    const plus_di = latest.PLUS_DI; const minus_di = latest.MINUS_DI;
    const di_spread = plus_di - minus_di;
    if(plus_di > minus_di){ if(di_spread >= 12) score += 15; else if(di_spread >= 8) score += 11; else if(di_spread >=4) score +=7; else score +=3 }

    const adxv = latest.ADX; const adx_rising = adxv > (adx.length>1?adx[adx.length-2].adx||adx[adx.length-2]:0);
    if(riskOff){
      if(di_spread < 6 || !adx_rising) score -= 8;
      if(adxv >= 35 && adx_rising) score += 12;
      else if(adxv >= 30 && adx_rising) score += 9;
      else if(adxv >=25 && adx_rising) score += 6;
      else score -= 8;
    }else{
      if(adxv >= 35 && adx_rising) score += 12; else if(adxv >= 30) score += 9; else if(adxv >=25) score += 6; else if(adxv >=20) score += 3;
    }

    if(higher_high && higher_close) score += 10; else if(consecutive_up) score += 6; else if(latest.Close > prev.Close) score +=3;

    if(rs >= 1.20) score += 8; else if(rs >= 1.10) score +=6; else if(rs >=1.03) score +=4; else if(rs >=0.98) score +=1;

    const rsiVal = latest.RSI;
    if(riskOff){
      if(55 <= rsiVal && rsiVal <= 68) score += 10;
      else if(rsiVal > 68 && rsiVal <= 75) score += 2;
      else if(rsiVal > 75) score -= 8;
      else if(rsiVal < 45) score -= 8;
    }else{
      if(55 <= rsiVal && rsiVal <= 70) score += 15; else if(rsiVal > 70 && rsiVal <= 80) score += 5; else if(rsiVal > 80) score -= 2; else if(rsiVal < 40) score -= 6;
    }

    if(riskOff){
      if(latest.MACD_hist > 0) score += 2;
      else score -= 3;
    }else if(latest.MACD_hist > 0) score += 2; // simplified

    if(!options_ok) score -= 4;

    const entry = latest.Close; const stop = entry - 1.2 * latest.ATR; const target = entry + 2.5 * latest.ATR;
    const risk_reward = (target-entry)/(entry-stop > 0 ? (entry-stop) : 1);

    return {
      symbol,
      score: Math.round(score*10)/10,
      close: Math.round(entry*100)/100,
      pct_from_high: Math.round(pct_from_high*100)/100,
      atr_from_high: Math.round(atr_from_high*100)/100,
      rel_vol_today: Math.round(rel_vol_today*100)/100,
      rel_vol_5d: Math.round(rel_vol_5d*100)/100,
      rsi: Math.round(rsiVal*10)/10,
      adx: Math.round(adxv*10)/10,
      plus_di: Math.round(plus_di*10)/10,
      rs_3m: Math.round(rs*100)/100,
      atr: Math.round(latest.ATR*100)/100,
      stop: Math.round(stop*100)/100,
      target: Math.round(target*100)/100,
      risk_reward: Math.round(risk_reward*100)/100,
      dollar_vol_20d: Math.round(dollarVol),
      options_ok: Boolean(options_ok),
      higher_high: Boolean(higher_high),
      consecutive_up: Boolean(consecutive_up)
    };
  }catch(e){ console.error(`  [!] ${symbol}: ${e.message}`); return null; }
}

async function main(){
  const cfg = readConfig();
  const riskMode = normalizeRiskMode(cfg);
  const tickerStr = cfg.TICKERS// || cfg.TICKERS || '';
  const tickers = tickerStr.split(/\s+/).filter(Boolean);
  if(!tickers.length){ console.error('No tickers in config (TEST or TICKERS)'); process.exit(1); }
  const start = new Date(); start.setDate(start.getDate() - 365);
  const startStr = formatDate(start);
  console.log(`Scoring ${tickers.length} tickers (mode=${riskMode})...`);
  const results = [];
  for(let i=0; i<tickers.length; i++){
    const s = tickers[i];
    process.stdout.write(`[${i+1}/${tickers.length}] ${s}...`);
    const r = await analyze(s, startStr, riskMode);
    if(r){ results.push(r); console.log(` score=${r.score}`); }
    else{ console.log(' skipped'); }
  }
  results.sort((a,b)=>b.score - a.score);
  const top40 = results.slice(0,40);
  const honorableMentions = top40.slice(20,40);
  const outFile = path.join(root, `momentum_results_${formatDate(new Date())}.json`);
  await fs.writeFile(outFile, JSON.stringify(top40, null, 2), 'utf8');
  console.log('Wrote', outFile);
  const honorableMentionsFile = path.join(root, `momentum_honorable_mentions_${formatDate(new Date())}.json`);
  await fs.writeFile(honorableMentionsFile, JSON.stringify(honorableMentions, null, 2), 'utf8');
  console.log('Wrote', honorableMentionsFile);
  const honorableMentionsLatestFile = path.join(root, 'momentum_honorable_mentions.json');
  await fs.writeFile(honorableMentionsLatestFile, JSON.stringify(honorableMentions, null, 2), 'utf8');
  console.log('Wrote', honorableMentionsLatestFile);
}

main().catch(e=>{ console.error(e); process.exit(1); });
